### PR TITLE
Add all basic layer types

### PR DESCRIPTION
**Pull request by wmatray**
- Add dropout, relu, reshapes
- Edit graph functions to work with evolution.py

Requires #4

### CODE REVIEW

**Review by PatrickNercessian — CHANGES_REQUESTED** [review]
Looks good! Minor changes mostly or just my curiosity, as usual.

Also my brain is fried so I haven't reviewed the rest of adapt_connections changes, but wanted to submit this review anyway

File: src/evolution.py [hunk 1]
```diff
@@ -69,7 +69,12 @@ def run_evolution(self, num_generations: int):
                 else:
                     child = self._mutate(copy.deepcopy(parent1))
                 child.id = self.id_counter
-                child.fitness = self.fitness_fn(child)
+                try:
+                    child.fitness = self.fitness_fn(child)
+                except Exception as e:
+                    print(f"Error in fitness function: {e}")
+                    child.fitness = 0
```
[the remark this answers is not in the document]
Comment (PatrickNercessian) [reply src/evolution.py:76]: We could either make it normal perplexity, and select for *lower* "fitness", or change this to some very negative number. Not sure if np -inf or whatever would work here or not since the other numbers are default python types

[the remark this answers is not in the document]
Comment (wmatray) [reply src/evolution.py:76]: Good catch, fixed

File: src/utils.py [hunk 2]
```diff
@@ -139,31 +170,134 @@ def add_skip_connection(graph, second_node, first_node, torch_function=torch.add
...
+    Helper function to adapt a tensor's size using repeat_interleave, circular padding, or adaptive pooling.
     
     Args:
         graph: The FX graph
-        node: The node whose shape needs to be adapted
-        current_size: Current size of the node's output
-        target_size: Desired size of the node's output
+        node: The node to adapt
+        current_size: Current size (integer)
+        target_size: Target size (integer)
```
[the remark this answers is not in the document]
Comment (wmatray) [reply src/utils.py:181]: Fixed

File: src/utils.py [hunk 3]
```diff
@@ -139,31 +170,134 @@ def add_skip_connection(graph, second_node, first_node, torch_function=torch.add
...
+        adapted_node: The node after adaptation
     """
-    if current_size == target_size:
-        return graph, node
-        
     if current_size < target_size:
-        # Need to increase size - use circular padding
-        graph, adapted_node = add_specific_node(graph, node, nn.CircularPad1d((0, target_size - current_size)))
+        length_multiplier = target_size // current_size
+        remainder = target_size % current_size
```
[the remark this answers is not in the document]
Comment (wmatray) [reply src/utils.py:189]: Sure, moved it

File: src/utils.py [hunk 4]
```diff
@@ -139,31 +170,134 @@ def add_skip_connection(graph, second_node, first_node, torch_function=torch.add
...
+        return graph, node
+    
+    if len(current_size) == 1:
+        # For 1D tensors, directly adapt the size
+        return adapt_tensor_size(graph, node, current_size[0], target_size[0], target_user)
+    
+    elif len(current_size) > 1:
+        # calculate total size of target shape by multiplying all dimensions except the first
+        target_total = math.prod(target_size)
+        current_total = math.prod(current_size)
```
[the remark this answers is not in the document]
Comment (wmatray) [reply src/utils.py:262]: Right, I'll make that more clear

File: src/utils.py [hunk 5]
```diff
@@ -139,31 +170,134 @@ def add_skip_connection(graph, second_node, first_node, torch_function=torch.add
...
+        # Adapt the flattened tensor
+        graph, adapted_node = adapt_tensor_size(
+            graph, 
+            flatten_node, 
+            current_total, 
+            target_total, 
+            target_user=target_user  # Intermediate node
+        )
+
+        # Add unflatten node
```
[the remark this answers is not in the document]
Comment (wmatray) [reply src/utils.py:291]: It's easiest to adapt dimensions in 1d then unflatten to the correct size

File: src/utils.py [hunk 6]
```diff
@@ -202,31 +336,28 @@ def add_branch_nodes(graph, reference_node, branch1_module, branch2_module):
         )
     
     # Run shape propagation to update metadata for the new nodes
-    example_input = torch.randn(reference_node_shape)
+    placeholder_shape = next(iter(graph.graph.nodes)).meta['tensor_meta'].shape
+    example_input = torch.randn(placeholder_shape)
```
[the remark this answers is not in the document]
Comment (wmatray) [reply src/utils.py:340]: Fixed

File: src/utils.py [hunk 7]
```diff
@@ -202,31 +336,28 @@ def add_branch_nodes(graph, reference_node, branch1_module, branch2_module):
...
     ShapeProp(graph).propagate(example_input)
     
     # Infer the shapes of the branch nodes from the metadata
-    branch1_shape = branch1_node.meta['tensor_meta'].shape
-    branch2_shape = branch2_node.meta['tensor_meta'].shape
+    branch1_shape = tuple(branch1_node.meta['tensor_meta'].shape[1:])
+    branch2_shape = tuple(branch2_node.meta['tensor_meta'].shape[1:])
```
[the remark this answers is not in the document]
Comment (wmatray) [reply src/utils.py:345]: done

File: src/core.py [hunk 8]
```diff
@@ -49,54 +49,97 @@ def add_node(graph: torch.fx.GraphModule, reference_node: torch.fx.Node, operati
...
+        input_size = kwargs.get('input_size')
+        output_size = kwargs.get('output_size')
+        
+        # If not provided in kwargs, try to use the ref_feature_shape or default to random
+        if input_size is None:
+            if len(ref_feature_shape) > 0:
+                input_size = ref_feature_shape[-1]
+            else:
+                input_size = random.randint(1, 1000)
+                print("Warning: Using random input_size for linear layer")
```
[the remark this answers is not in the document]
Comment (wmatray) [reply src/core.py:79]: Yes you're right, unnecessary

File: src/core.py [hunk 9]
```diff
@@ -49,54 +49,97 @@ def add_node(graph: torch.fx.GraphModule, reference_node: torch.fx.Node, operati
...
             - For 'pool': target_size
             - For 'repeat': target_size
+            - For 'skip': first_node
     Returns:
         graph: The modified graph
     """
     
     # Get required shapes before making any modifications
-    input_shape, output_shape = find_required_shapes(reference_node)
+    ref_input_shape, ref_output_shape = find_required_shapes(reference_node)  # SHAPE NOTE: Returns shapes with batch dimension included
```
[the remark this answers is not in the document]
Comment (wmatray) [reply src/core.py:62]: The whole line is actually redundant, removed

File: src/core.py [hunk 10]
```diff
@@ -105,56 +148,110 @@ def add_node(graph: torch.fx.GraphModule, reference_node: torch.fx.Node, operati
...
+        # Get feature dimensions from skip connection output shape
+        new_node_output_shape = get_feature_dims(skip_output_shape)
+        
+        # For branches, input shape is reference node features, output is from skip connection
+        new_node_input_shape = ref_feature_shape
+
+    # Add a dropout layer
+    elif operation == 'dropout':
+        # Get dropout probability from kwargs if provided, otherwise use default
+        prob = kwargs.get('prob', 0.5)
```
[the remark this answers is not in the document]
Comment (wmatray) [reply src/core.py:193]: I don't think 0.5 is crazy, but I thought core.py and lower level functions are only going to sample for hyperparameters when absolutely necessary like with branching. If we want to choose dropout rate from a distribution it will happen in our mutation functions. Or are the mutations in core.py fully loaded and we call them directly?

File: src/core.py [hunk 11]
```diff
@@ -105,56 +148,110 @@ def add_node(graph: torch.fx.GraphModule, reference_node: torch.fx.Node, operati
...
+    # TODO: Implement different removals for skip connections and branch nodes
+
+
     input_node = reference_node.args[0]
     
     # Get shapes before removing node
-    output_left_shape = list(input_node.meta['tensor_meta'].shape)
-    input_right_shape = reference_node.meta['tensor_meta'].shape
+    output_left_shape = list(input_node.meta['tensor_meta'].shape)  # SHAPE NOTE: Full shape with batch dimension
+    input_right_shape = reference_node.meta['tensor_meta'].shape  # SHAPE NOTE: Full shape with batch dimension
```
[the remark this answers is not in the document]
Comment (wmatray) [reply src/core.py:251]: There's literally no reason, fixed

File: src/core.py [hunk 12]
```diff
@@ -163,76 +260,83 @@ def remove_node(graph: torch.fx.GraphModule, reference_node: torch.fx.Node):
...
+    child_features = get_feature_dims(child_input_shape)
+    
     # Special handling for skip connections (torch.add operations)
+    # TODO: Handle any kind of skip connection (e.g. torch.cat, torch.mul, etc.)
     if new_node.target == torch.add:
         # Get shapes of both input nodes
-        first_node = new_node.args[0]
-        second_node = new_node.args[1]
+        first_node = new_node.args[1]
+        second_node = new_node.args[0]
```
[the remark this answers is not in the document]
Comment (wmatray) [reply src/core.py:311]: That was done because of the way I had assigned args in the skip and branch functions. Just committed a fix to be less confusing

File: src/utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (PatrickNercessian) [reply src/utils.py:292]: Does this mean we'll just have loads of flatten/unflatten nodes everywhere? Is that bad, is there significant overhead to having more nodes in general, or to having those nodes specifically?

File: src/core.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (PatrickNercessian) [reply src/core.py:186]: That's kinda what I'm saying above:

> I guess the calling function could do the random gen, but that's inconsistent with at least one other place in this function (the random output layer dimension for 'linear' operations)

So if we want the distribution sampling to happen in the mutation functions, then we probably need to make that consistent (e.g. I saw the linear layer dim size was sampled in core.py I think, and maybe there are others)

File: src/utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (wmatray) [reply src/utils.py:292]: Flatten/unflatten is like a view and negligible, padding and pooling will introduce some overhead but not as much as a trainable node like linear or convolutional. Pooling also creates information loss.

[the remark this answers is not in the document]
Comment (wmatray) [reply src/utils.py:292]: The fitness function will make sure our model is effective even if there's a ton of pooling and padding nodes. The bad models won't make it.

Here's an example of the top architecture several generations on a weather dataset with mutation rate at 100%:
image

File: src/core.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (wmatray) [reply src/core.py]: I agree, I removed the cases where that happens. It's up to the evo function to determine that.

**Review by PatrickNercessian — APPROVED** [review]
Nice work!

File: src/utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (MortimerE) [reply src/utils.py:292]: Correct me if I'm simply misreading, but couldn't this cause dimensional alignment issues with the input? Flattening creates a single contiguous dimension, and reshaping a matrix from the flat vector would result in different dimensional relationships within a tensor. Two results of this that might cause issues would be concatenation of previously independent segments into the same dimension or separation of previously contiguous segments into different dimensions. My feeling is that this might be along the same lines of concerns with locality elsewhere in the project.
e.g. 
[1, 2][3, 4] -> [1, 2, 3, 4] -> [1, 2, 3][4, 0, 0] 
In this case, 3 and 4 were originally elements of the same dimension but 3 is now associated dimensionally with 1, 2 and not with 4. It might make more sense to do a strided padding/interleaving here and get something like [1, 2, 0][3, 4, 0] 
This example doesn't really cover the cases where we need to shrink the matrix or when we shrink just one dimension and expand another, but this seemed like something that was worth discussing further.

File: src/core.py [hunk 13]
```diff
@@ -49,54 +49,89 @@ def add_node(graph: torch.fx.GraphModule, reference_node: torch.fx.Node, operati
...
+        # Input and output shapes for repeat
+        new_node_input_shape = ref_feature_shape
+        # Create a new tuple with the last dimension changed
+        if len(ref_feature_shape) == 1:
+            new_node_output_shape = (target_size,)
+        else:
+            new_node_output_shape = tuple(list(ref_feature_shape[:-1]) + [target_size])
 
     # Add a flatten layer, that flattens every dimension except the batch dimension 
     elif operation == 'flatten':
```
Comment (MortimerE) [inline src/core.py:126]: Suggestion: add an "expansion" operation as inverse flattening, where individual vectors are broadcast into higher dimensional shapes. The original transformer decoder architecture does this to allow sort of an annealing property after applying attention blocks 

This would just be a reshape operation and wouldn't require params except for maybe a learnable ratio of dimensionality expansion, so like flattening it would be low overhead

**Review by MortimerE — COMMENTED** [review]
Added a few comments regarding dimensionality of tensors to consider for a future PR

File: src/utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (wmatray) [reply src/utils.py:292]: Let's talk about this in the call with some whiteboarding. I'm a little confused by what you're getting at

File: src/core.py (hunk 13 above)
Comment (wmatray) [reply src/core.py:126]: The problem with adding a reshape by itself is that the adaptation framework would immediately add another reshape that would undo the process. This one of the issues with our current system. The alternative is to use just graphs and then assign shapes to the whole network after the fact.

Comment (MortimerE) [reply src/core.py:126]: That's more or less what happens in the decoder anyway, there's a reshape, a linear layer, and then a re-embedding

File: src/utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (PatrickNercessian) [reply src/utils.py:292]: Screenshot 2025-05-20 at 12 01 57 PM
